<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Bioconductor and the SingleCellExperiment class • ISMB.OSCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Bioconductor and the SingleCellExperiment class">
<meta property="og:description" content="ISMB.OSCA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISMB.OSCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_bioc_sce.html">Introduction to Bioconductor</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Bioconductor/ISMB.OSCA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Bioconductor and the
SingleCellExperiment class</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/ISMB.OSCA/blob/HEAD/vignettes/intro_bioc_sce.Rmd" class="external-link"><code>vignettes/intro_bioc_sce.Rmd</code></a></small>
      <div class="hidden name"><code>intro_bioc_sce.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/MouseGastrulationData" class="external-link">MouseGastrulationData</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-singlecellexperiment-class">The <code>SingleCellExperiment</code> class<a class="anchor" aria-label="anchor" href="#the-singlecellexperiment-class"></a>
</h2>
<p>One of the main strengths of the Bioconductor project lies in the use
of a common data infrastructure that powers interoperability across
packages.</p>
<p>Users should be able to analyze their data using functions from
different Bioconductor packages without the need to convert between
formats. To this end, the <code>SingleCellExperiment</code> class (from
the <em>SingleCellExperiment</em> package) serves as the common currency
for data exchange across 70+ single-cell-related Bioconductor
packages.</p>
<p>This class implements a data structure that stores all aspects of our
single-cell data - gene-by-cell expression data, per-cell metadata and
per-gene annotation - and manipulate them in a synchronized manner.</p>
<p><img src="http://bioconductor.org/books/3.17/OSCA.intro/images/SingleCellExperiment.png"><!-- --></p>
<p>Let’s start with an example dataset.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/WTChimeraData.html" class="external-link">WTChimeraData</a></span><span class="op">(</span>sample<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 29453 2411 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">##   ENSMUSG00000095742 tomato-td</span></span>
<span><span class="co">## rowData names(2): ENSEMBL SYMBOL</span></span>
<span><span class="co">## colnames(2411): cell_9769 cell_9770 ... cell_12178 cell_12179</span></span>
<span><span class="co">## colData names(11): cell barcode ... doub.density sizeFactor</span></span>
<span><span class="co">## reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>We can think of this (and other) class as a <em>container</em>, that
contains several different pieces of data in so-called
<em>slots</em>.</p>
<p>The <em>getter</em> methods are used to extract information from the
slots and the <em>setter</em> methods are used to add information into
the slots. These are the only ways to interact with the objects (rather
than directly accessing the slots).</p>
<p>Depending on the object, slots can contain different types of data
(e.g., numeric matrices, lists, etc.). We will here review the main
slots of the SingleCellExperiment class as well as their getter/setter
methods.</p>
<div class="section level3">
<h3 id="the-assays">The <code>assays</code><a class="anchor" aria-label="anchor" href="#the-assays"></a>
</h3>
<p>This is arguably the most fundamental part of the object that
contains the count matrix, and potentially other matrices with
transformed data. We can access the <em>list</em> of matrices with the
<code>assays</code> function and individual matrices with the
<code>assay</code> function. If one of these matrices is called
“counts”, we can use the special <code>counts</code> getter (and the
analogous <code>logcounts</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 3 x 3 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##                    cell_9769 cell_9770 cell_9771</span></span>
<span><span class="co">## ENSMUSG00000051951         .         .         .</span></span>
<span><span class="co">## ENSMUSG00000089699         .         .         .</span></span>
<span><span class="co">## ENSMUSG00000102343         .         .         .</span></span></code></pre>
<p>You will notice that in this case we have a sparse matrix of class
“dgTMatrix” inside the object. More generally, any “matrix-like” object
can be used, e.g., dense matrices or HDF5-backed matrices (see “Working
with large data”).</p>
</div>
<div class="section level3">
<h3 id="the-coldata-and-rowdata">The <code>colData</code> and <code>rowData</code><a class="anchor" aria-label="anchor" href="#the-coldata-and-rowdata"></a>
</h3>
<p>Conceptually, these are two data frames that annotate the columns and
the rows of your assay, respectively.</p>
<p>One can interact with them as usual, e.g., by extracting columns or
adding additional variables as columns.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2411 rows and 11 columns</span></span>
<span><span class="co">##                   cell          barcode    sample       stage    tomato</span></span>
<span><span class="co">##            &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;logical&gt;</span></span>
<span><span class="co">## cell_9769    cell_9769 AAACCTGAGACTGTAA         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9770    cell_9770 AAACCTGAGATGCCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9771    cell_9771 AAACCTGAGCAGCCTC         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9772    cell_9772 AAACCTGCATACTCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9773    cell_9773 AAACGGGTCAACACCA         5        E8.5      TRUE</span></span>
<span><span class="co">## ...                ...              ...       ...         ...       ...</span></span>
<span><span class="co">## cell_12175  cell_12175 TTTGGTTAGTCCGTAT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12176  cell_12176 TTTGGTTAGTGTTGAA         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12177  cell_12177 TTTGGTTGTTAAAGAC         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12178  cell_12178 TTTGGTTTCAGTCAGT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12179  cell_12179 TTTGGTTTCGCCATAA         5        E8.5      TRUE</span></span>
<span><span class="co">##                 pool stage.mapped        celltype.mapped closest.cell</span></span>
<span><span class="co">##            &lt;integer&gt;  &lt;character&gt;            &lt;character&gt;  &lt;character&gt;</span></span>
<span><span class="co">## cell_9769          3        E8.25             Mesenchyme   cell_24159</span></span>
<span><span class="co">## cell_9770          3         E8.5            Endothelium   cell_96660</span></span>
<span><span class="co">## cell_9771          3         E8.5              Allantois  cell_134982</span></span>
<span><span class="co">## cell_9772          3         E8.5             Erythroid3  cell_133892</span></span>
<span><span class="co">## cell_9773          3        E8.25             Erythroid1   cell_76296</span></span>
<span><span class="co">## ...              ...          ...                    ...          ...</span></span>
<span><span class="co">## cell_12175         3         E8.5             Erythroid3  cell_138060</span></span>
<span><span class="co">## cell_12176         3         E8.5 Forebrain/Midbrain/H..   cell_72709</span></span>
<span><span class="co">## cell_12177         3        E8.25       Surface ectoderm  cell_100275</span></span>
<span><span class="co">## cell_12178         3        E8.25             Erythroid2   cell_70906</span></span>
<span><span class="co">## cell_12179         3         E8.5            Spinal cord  cell_102334</span></span>
<span><span class="co">##            doub.density sizeFactor</span></span>
<span><span class="co">##               &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## cell_9769    0.02985045    1.41243</span></span>
<span><span class="co">## cell_9770    0.00172753    1.22757</span></span>
<span><span class="co">## cell_9771    0.01338013    1.15439</span></span>
<span><span class="co">## cell_9772    0.00218402    1.28676</span></span>
<span><span class="co">## cell_9773    0.00211723    1.78719</span></span>
<span><span class="co">## ...                 ...        ...</span></span>
<span><span class="co">## cell_12175   0.00129403   1.219506</span></span>
<span><span class="co">## cell_12176   0.01833074   1.095753</span></span>
<span><span class="co">## cell_12177   0.03104037   0.910728</span></span>
<span><span class="co">## cell_12178   0.00169483   2.061701</span></span>
<span><span class="co">## cell_12179   0.03767894   1.798687</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 29453 rows and 2 columns</span></span>
<span><span class="co">##                               ENSEMBL         SYMBOL</span></span>
<span><span class="co">##                           &lt;character&gt;    &lt;character&gt;</span></span>
<span><span class="co">## ENSMUSG00000051951 ENSMUSG00000051951           Xkr4</span></span>
<span><span class="co">## ENSMUSG00000089699 ENSMUSG00000089699         Gm1992</span></span>
<span><span class="co">## ENSMUSG00000102343 ENSMUSG00000102343        Gm37381</span></span>
<span><span class="co">## ENSMUSG00000025900 ENSMUSG00000025900            Rp1</span></span>
<span><span class="co">## ENSMUSG00000025902 ENSMUSG00000025902          Sox17</span></span>
<span><span class="co">## ...                               ...            ...</span></span>
<span><span class="co">## ENSMUSG00000095041 ENSMUSG00000095041     AC149090.1</span></span>
<span><span class="co">## ENSMUSG00000063897 ENSMUSG00000063897          DHRSX</span></span>
<span><span class="co">## ENSMUSG00000096730 ENSMUSG00000096730       Vmn2r122</span></span>
<span><span class="co">## ENSMUSG00000095742 ENSMUSG00000095742 CAAA01147332.1</span></span>
<span><span class="co">## tomato-td                   tomato-td      tomato-td</span></span></code></pre>
<p>Note the <code>$</code> short cut.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">sum</span>, <span class="va">sce</span><span class="op">$</span><span class="va">sum</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">my_sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2411 rows and 12 columns</span></span>
<span><span class="co">##                   cell          barcode    sample       stage    tomato</span></span>
<span><span class="co">##            &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;logical&gt;</span></span>
<span><span class="co">## cell_9769    cell_9769 AAACCTGAGACTGTAA         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9770    cell_9770 AAACCTGAGATGCCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9771    cell_9771 AAACCTGAGCAGCCTC         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9772    cell_9772 AAACCTGCATACTCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9773    cell_9773 AAACGGGTCAACACCA         5        E8.5      TRUE</span></span>
<span><span class="co">## ...                ...              ...       ...         ...       ...</span></span>
<span><span class="co">## cell_12175  cell_12175 TTTGGTTAGTCCGTAT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12176  cell_12176 TTTGGTTAGTGTTGAA         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12177  cell_12177 TTTGGTTGTTAAAGAC         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12178  cell_12178 TTTGGTTTCAGTCAGT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_12179  cell_12179 TTTGGTTTCGCCATAA         5        E8.5      TRUE</span></span>
<span><span class="co">##                 pool stage.mapped        celltype.mapped closest.cell</span></span>
<span><span class="co">##            &lt;integer&gt;  &lt;character&gt;            &lt;character&gt;  &lt;character&gt;</span></span>
<span><span class="co">## cell_9769          3        E8.25             Mesenchyme   cell_24159</span></span>
<span><span class="co">## cell_9770          3         E8.5            Endothelium   cell_96660</span></span>
<span><span class="co">## cell_9771          3         E8.5              Allantois  cell_134982</span></span>
<span><span class="co">## cell_9772          3         E8.5             Erythroid3  cell_133892</span></span>
<span><span class="co">## cell_9773          3        E8.25             Erythroid1   cell_76296</span></span>
<span><span class="co">## ...              ...          ...                    ...          ...</span></span>
<span><span class="co">## cell_12175         3         E8.5             Erythroid3  cell_138060</span></span>
<span><span class="co">## cell_12176         3         E8.5 Forebrain/Midbrain/H..   cell_72709</span></span>
<span><span class="co">## cell_12177         3        E8.25       Surface ectoderm  cell_100275</span></span>
<span><span class="co">## cell_12178         3        E8.25             Erythroid2   cell_70906</span></span>
<span><span class="co">## cell_12179         3         E8.5            Spinal cord  cell_102334</span></span>
<span><span class="co">##            doub.density sizeFactor    my_sum</span></span>
<span><span class="co">##               &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## cell_9769    0.02985045    1.41243     27577</span></span>
<span><span class="co">## cell_9770    0.00172753    1.22757     29309</span></span>
<span><span class="co">## cell_9771    0.01338013    1.15439     28795</span></span>
<span><span class="co">## cell_9772    0.00218402    1.28676     34794</span></span>
<span><span class="co">## cell_9773    0.00211723    1.78719     38300</span></span>
<span><span class="co">## ...                 ...        ...       ...</span></span>
<span><span class="co">## cell_12175   0.00129403   1.219506     26680</span></span>
<span><span class="co">## cell_12176   0.01833074   1.095753     19013</span></span>
<span><span class="co">## cell_12177   0.03104037   0.910728     24627</span></span>
<span><span class="co">## cell_12178   0.00169483   2.061701     46162</span></span>
<span><span class="co">## cell_12179   0.03767894   1.798687     38398</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="the-reduceddims">The <code>reducedDims</code><a class="anchor" aria-label="anchor" href="#the-reduceddims"></a>
</h3>
<p>Everything that we have described so far (except for the
<code>counts</code> getter) is part of the
<code>SummarizedExperiment</code> class that SingleCellExperiment
extends.</p>
<p>One of the peculiarity of SingleCellExperiment is its ability to
store reduced dimension matrices within the object. These may include
PCA, t-SNE, UMAP, etc.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 2</span></span>
<span><span class="co">## names(2): pca.corrected.E7.5 pca.corrected.E8.5</span></span></code></pre>
<p>As for the other slots, we have the usual setter/getter, but it is
somewhat rare to interact directly with these functions.</p>
<p>It is more common for other functions to <em>store</em> this
information in the object, e.g., the <code>runPCA</code> function from
the <code>scater</code> package.</p>
<p>Here, we use <code>scater</code>’s <code>plotReducedDim</code>
function as an example of how to extract this information
<em>indirectly</em> from the objects. Note that one could obtain the
same results (somewhat less efficiently) by extracting the corresponding
<code>reducedDim</code> matrix and <code>ggplot</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: scuttle</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"pca.corrected.E8.5"</span>, colour_by <span class="op">=</span> <span class="st">"celltype.mapped"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 131 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_point()`).</span></span></code></pre>
<p><img src="intro_bioc_sce_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] scater_1.33.1                ggplot2_3.5.1               </span></span>
<span><span class="co">##  [3] scuttle_1.15.0               MouseGastrulationData_1.19.0</span></span>
<span><span class="co">##  [5] SpatialExperiment_1.15.0     SingleCellExperiment_1.27.2 </span></span>
<span><span class="co">##  [7] SummarizedExperiment_1.35.0  Biobase_2.65.0              </span></span>
<span><span class="co">##  [9] GenomicRanges_1.57.1         GenomeInfoDb_1.41.1         </span></span>
<span><span class="co">## [11] IRanges_2.39.0               S4Vectors_0.43.0            </span></span>
<span><span class="co">## [13] BiocGenerics_0.51.0          MatrixGenerics_1.17.0       </span></span>
<span><span class="co">## [15] matrixStats_1.3.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] DBI_1.2.3                 gridExtra_2.3            </span></span>
<span><span class="co">##   [3] rlang_1.1.4               magrittr_2.0.3           </span></span>
<span><span class="co">##   [5] compiler_4.4.0            RSQLite_2.3.7            </span></span>
<span><span class="co">##   [7] DelayedMatrixStats_1.27.1 png_0.1-8                </span></span>
<span><span class="co">##   [9] systemfonts_1.1.0         vctrs_0.6.5              </span></span>
<span><span class="co">##  [11] pkgconfig_2.0.3           crayon_1.5.2             </span></span>
<span><span class="co">##  [13] fastmap_1.2.0             dbplyr_2.5.0             </span></span>
<span><span class="co">##  [15] magick_2.8.3              XVector_0.45.0           </span></span>
<span><span class="co">##  [17] labeling_0.4.3            utf8_1.2.4               </span></span>
<span><span class="co">##  [19] rmarkdown_2.27            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##  [21] UCSC.utils_1.1.0          ragg_1.3.2               </span></span>
<span><span class="co">##  [23] purrr_1.0.2               bit_4.0.5                </span></span>
<span><span class="co">##  [25] xfun_0.44                 zlibbioc_1.51.1          </span></span>
<span><span class="co">##  [27] cachem_1.1.0              beachmat_2.21.3          </span></span>
<span><span class="co">##  [29] jsonlite_1.8.8            blob_1.2.4               </span></span>
<span><span class="co">##  [31] highr_0.11                DelayedArray_0.31.1      </span></span>
<span><span class="co">##  [33] BiocParallel_1.39.0       irlba_2.3.5.1            </span></span>
<span><span class="co">##  [35] parallel_4.4.0            R6_2.5.1                 </span></span>
<span><span class="co">##  [37] bslib_0.7.0               jquerylib_0.1.4          </span></span>
<span><span class="co">##  [39] Rcpp_1.0.12               knitr_1.47               </span></span>
<span><span class="co">##  [41] Matrix_1.7-0              tidyselect_1.2.1         </span></span>
<span><span class="co">##  [43] viridis_0.6.5             abind_1.4-5              </span></span>
<span><span class="co">##  [45] yaml_2.3.8                codetools_0.2-20         </span></span>
<span><span class="co">##  [47] curl_5.2.1                lattice_0.22-6           </span></span>
<span><span class="co">##  [49] tibble_3.2.1              withr_3.0.0              </span></span>
<span><span class="co">##  [51] KEGGREST_1.45.0           BumpyMatrix_1.13.0       </span></span>
<span><span class="co">##  [53] evaluate_0.24.0           desc_1.4.3               </span></span>
<span><span class="co">##  [55] BiocFileCache_2.13.0      ExperimentHub_2.13.0     </span></span>
<span><span class="co">##  [57] Biostrings_2.73.1         pillar_1.9.0             </span></span>
<span><span class="co">##  [59] BiocManager_1.30.23       filelock_1.0.3           </span></span>
<span><span class="co">##  [61] generics_0.1.3            BiocVersion_3.20.0       </span></span>
<span><span class="co">##  [63] sparseMatrixStats_1.17.2  munsell_0.5.1            </span></span>
<span><span class="co">##  [65] scales_1.3.0              glue_1.7.0               </span></span>
<span><span class="co">##  [67] tools_4.4.0               AnnotationHub_3.13.0     </span></span>
<span><span class="co">##  [69] BiocNeighbors_1.23.0      ScaledMatrix_1.13.0      </span></span>
<span><span class="co">##  [71] fs_1.6.4                  cowplot_1.1.3            </span></span>
<span><span class="co">##  [73] grid_4.4.0                AnnotationDbi_1.67.0     </span></span>
<span><span class="co">##  [75] colorspace_2.1-0          GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">##  [77] beeswarm_0.4.0            BiocSingular_1.21.1      </span></span>
<span><span class="co">##  [79] vipor_0.4.7               rsvd_1.0.5               </span></span>
<span><span class="co">##  [81] cli_3.6.2                 rappdirs_0.3.3           </span></span>
<span><span class="co">##  [83] textshaping_0.4.0         fansi_1.0.6              </span></span>
<span><span class="co">##  [85] viridisLite_0.4.2         S4Arrays_1.5.1           </span></span>
<span><span class="co">##  [87] dplyr_1.1.4               gtable_0.3.5             </span></span>
<span><span class="co">##  [89] sass_0.4.9                digest_0.6.35            </span></span>
<span><span class="co">##  [91] ggrepel_0.9.5             SparseArray_1.5.7        </span></span>
<span><span class="co">##  [93] farver_2.1.2              rjson_0.2.21             </span></span>
<span><span class="co">##  [95] htmlwidgets_1.6.4         memoise_2.0.1            </span></span>
<span><span class="co">##  [97] htmltools_0.5.8.1         pkgdown_2.0.9            </span></span>
<span><span class="co">##  [99] lifecycle_1.0.4           httr_1.4.7               </span></span>
<span><span class="co">## [101] mime_0.12                 bit64_4.0.5</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="further-reading">Further Reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h2>
<ul>
<li>OSCA book, <a href="https://bioconductor.org/books/release/OSCA.intro" class="external-link">Introduction</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Create a <code>SingleCellExperiment</code> object: Try and create a
SingleCellExperiment object “from scratch”. Start from a matrix (either
randomly generated or with some fake data in it) and add one or more
columns as colData.</li>
</ol>
<p>Hint: the <code>SingleCellExperiment</code> function can be used to
create a new SingleCellExperiment object.</p>
<ol start="2" style="list-style-type: decimal">
<li>Combining two objects: The <code>MouseGastrulationData</code>
package contains several datasets. Download sample 6 of the chimera
experiment by running <code>sce6 &lt;- WTChimeraData(sample=6)</code>.
Use the <code>cbind</code> function to combine the new data with the
<code>sce</code> object created before.</li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Davide Risso, Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
